### PR TITLE
First pass at refactoring

### PR DESCRIPTION
**Pull request by ksin**
Didn't add any features. Just made it easier for me to read. Made a couple helper functions and DRYed things up a bit. Mostly groundwork. Changes:
- Some computed properties added to Building model. (see end of that file)
- Reformatting (sorry for the ugly diff).
- Clarifying the logic of a few functions in game.js
- Remove unused code and excessive comments.
- Set some values as constants.
- Encapsulate start of app with View.init function.
- Extract common functions: `wereUnitsDestroyed` and `statsForDestroyed`

Things to note:
- Try to use better function/variable names before resorting to comments. Most of the logic is pretty easy to follow.
- Separate getter functions, setter functions, and functions that perform work. Sometimes one function does too much. Consequence is there are functions a reader thinks is a getter, but in reality there's some side effect. (ie: [`Game#areBuildingsDestroyed` sets an `undefined` when it's just supposed to retrieve a boolean](https://github.com/tsamb/rts-tower-defence/pull/18#discussion-diff-28177825)).
- Already noted in your comments: extract modules and mixins where possible. 
- If you want to be cool, use ES6 module imports to not pollute global namespace. Also makes it very explicit what modules are being used. [es6to5](http://es6rocks.com/2014/10/es6-modules-today-with-6to5/)

The Game model is kind of bulky. There are several things I think you can abstract out of it that'll make the code easier to read. Here are some suggestions:
- `this.buildings` can be some sort of BuildingCollection model. Lots of functions that are isolated to building collections: `fireBuildingsAtEnemies`, `buildCurrentBuildOrder`, `wereUnitsDestroyed`, `statsForDestroyed`, `buildingProducedResources`, etc...
- same for `this.enemies` 
- There are lots of similar methods between building collections and enemy collections. Basically, anything having to do with updating the DOM (ie: `needsUpdate`). Can probably use some mixin strategy.
- Instead of `BuildingSelector` how about just a `Builder` that manages all that plus the resources.
- A ResourcesManager might be a good abstraction too since there are several functions related to the management of matter and energy (ie: `updateResources`).

Let me know what you think, and if anything broke! There aren't any tests and you know how the code should work better than me.

### CODE REVIEW

File: js/game.js [hunk 1]
```diff
@@ -206,92 +187,67 @@ Game.prototype.build = function(xOnBoard,yOnBoard) {
...
+};
 
-Game.prototype.areBuildingsDestroyed = function() {
-  var isAtLeastOneDestroyed = false
+Game.prototype.fireBuildingsAtEnemies = function() {
   for (var i = 0; i < this.buildings.length; i++) {
-    if (this.buildings[i].isDestroyed()) {
-      isAtLeastOneDestroyed = true;
-      if (!this.buildings[i].completed) {
-        this.currentBuildOrder = undefined;
```
Comment (ksin) [inline js/game.js:230]: In the original function, this is a side effect that's unexpected from reading the function's name. I just removed it for now since it doesn't seem to break anything. Thoughts?

File: js/game.js [hunk 2]
```diff
@@ -206,92 +187,67 @@ Game.prototype.build = function(xOnBoard,yOnBoard) {
...
     View.deselectBuilding();
   }
-}
+};
 
 Game.prototype.buildCurrentBuildOrder = function() {
-  var building = this.currentBuildOrder
-  if (building) {
-    if (building.completed) {
-      building = undefined;
```
Comment (ksin) [inline js/game.js:215]: I don't think this does anything. You're setting the variable to `undefined`, not the model property. The variable is never returned. I just removed the else statement.

Comment (ksin) [inline js/game.js]: I don't think this does anything. You're setting the variable to `undefined`, not the model property. The variable is never returned. Just removed the else statement.

Comment (ksin) [inline js/game.js]: I don't think this does anything. You're setting the variable to `undefined`, not the model property. The variable is never returned. Just removed the else statement.

File: js/invariant-data.js [hunk 3]
```diff
@@ -1,135 +1,140 @@
...
+    key: "s"
+  }
+];
 
 var GameOptions = {
   STARTING_MATTER: 1000,
   STARTING_ENERGY: 5000,
   COMMAND_CENTER: new Building(BuildingsList[0])
-}
+};
```
Comment (tsamb) [inline js/invariant-data.js:101]: Wait... was this whole massive red/green diff caused by this one semicolon?

Comment (ksin) [inline js/invariant-data.js:101]: git isn't perfect :disappointed:

**Comment by tsamb** [comment]
Thanks for working through this; I'll walk through it later. It'd be great if you could extract some of your PR suggestions into issues on the repo!

**Comment by tsamb** [comment]
@ksin can you please amend your commit message to be slightly more semantic about what you refactored?

**Comment by ksin** [comment]
@tsamb how's that? A little verbose. But descriptive!

File: js/view.js [hunk 4]
```diff
@@ -1,114 +1,114 @@
-// View
-
 var View = (function() {
-  View = {};
+  return {
```
Comment (tsamb) [inline js/view.js:2]: Why this change? Too much repetition of `View`? I kept it the other way because I find the `class.functionName =` syntax way easier to read. I like to be able to scan the file and say "ah yes, `View` has a `setBuildListeners` method on it..."

It's a personal preference of mine. I find the object literal notation harder to read with the semicolons.

File: js/application.js [hunk 5]
```diff
@@ -1,8 +1,6 @@
 var game; // for development to access game object in browser
 
-// Game logic
 $(document).ready(function() {
-  View.prependBuildingButtons(BuildingsList)
-  View.enablePauseButton();
-  game = new Game;
+  View.init();
```
Comment (tsamb) [inline js/application.js:4]: Do you think we should inject the `BuildingsList` dependency here, rather than referencing it within the method in the View module?

File: js/building.js [hunk 6]
```diff
@@ -1,7 +1,7 @@
-// Building model
+var FRAMES_PER_RUN = 5; // how often (in frames) the build loop runs
```
Comment (tsamb) [inline js/building.js:1]: I feel icky about this being here.

File: js/game.js [hunk 7]
```diff
@@ -206,92 +187,67 @@ Game.prototype.build = function(xOnBoard,yOnBoard) {
...
-    }
-  }
-  return isAtLeastOneDestroyed;
-}
-
-Game.prototype.buildingsFire = function() {
-  for (var i = 0; i < this.buildings.length; i++) {
-    // if it has a position.x it must already be placed
-    if (this.buildings[i].damagePerShot && this.buildings[i].position.x) {
+    if (this.buildings[i].canFire()) {
```
Comment (tsamb) [inline js/game.js:203]: Nice refactoring.

File: js/building.js [hunk 8]
```diff
@@ -29,86 +29,97 @@ function Building(options, game) {
...
+      var damage = this.damagePerShot; // Math.floor(Math.random() * this.damagePerShot);
       this.game.deductEnergy(this.energyPerShot);
       this.game.board.drawLaser(this.centerX(), this.centerY(), closestEnemy.centerX(), closestEnemy.centerY());
       closestEnemy.receiveDamage(damage);
     }
   }
-}
+};
+
+///// Computed properties /////
```
Comment (tsamb) [inline js/building.js:115]: Nice addition.

File: js/enemy.js [hunk 9]
```diff
@@ -1,8 +1,8 @@
-// Enemy model
+var DEFAULT_ENEMY_SIZE = 10;
```
Comment (tsamb) [inline js/enemy.js:1]: Good. But now that we're polluting the global name spaces with variables, we should probably wrap this class (and the others) in an IIFE.

File: js/game.js [hunk 10]
```diff
@@ -206,92 +187,67 @@ Game.prototype.build = function(xOnBoard,yOnBoard) {
...
-Difficulty.enemyHp = function(level) {
-  // TKTKTK: where should this 50 base hp per level live?
-  return 50 * level;
-}
+    totalDamageDealt: totalDamageDealt,
+    numDestroyed: units.length
+  };
+};
+
+var Difficulty = {
```
Comment (tsamb) [inline js/game.js:238]: Raaargh! @ksin you and your inline object notation!

**Comment by tsamb** [comment]
Thanks for the changes, @ksin. Merging.
